src/snovault/elasticsearch/indexer.py

```diff
@@ -143,140 +154,302 @@ def get_related_uuids(request, es, updated, renamed):
+                updated = set()
+                renamed = set()
+                max_xid = 0
+                txn_count = 0
+                for txn in txns.all():
+                    txn_count += 1
+                    max_xid = max(max_xid, txn.xid)
+                    if first_txn is None:
+                        first_txn = txn.timestamp
+                    else:
+                        first_txn = min(first_txn, txn.timestamp)
+                    renamed.update(txn.data.get('renamed', ()))
+                    updated.update(txn.data.get('updated', ()))
+
+                if invalidated:        # reindex requested, treat like updated
+                    updated |= invalidated
+
+                result['txn_count'] = txn_count
+                if txn_count == 0 and len(invalidated) == 0:
+                    indexer_state.send_notices()
+                    return_now = True
+                    return result, return_now, flush, invalidated, first_txn, snapshot_id, restart, xmin
+
+                is_testing = asbool(request.registry.settings.get('testing', False))
+                is_testing_full = request.json.get('is_testing_full', False)
+                if is_testing and is_testing_full:
+                    full_reindex = False
+                    related_set = set(all_uuids(request.registry))
+                else:
+                    (related_set, full_reindex) = get_related_uuids(request, request.registry[ELASTIC_SEARCH], updated, renamed)
+                if full_reindex:
+                    invalidated = related_set
+                    flush = True
+                else:
+                    invalidated = related_set | updated
+                    result.update(
+                        max_xid=max_xid,
+                        renamed=renamed,
+                        updated=updated,
+                        referencing=len(related_set),
+                        invalidated=len(invalidated),
+                        txn_count=txn_count
+                    )
+                    if first_txn is not None:
+                        result['first_txn_timestamp'] = first_txn.isoformat()
+
+                if invalidated and not request.json.get('dry_run', False):
+                    # Exporting a snapshot mints a new xid, so only do so when required.
+                    # Not yet possible to export a snapshot on a standby server:
+                    # http://www.postgresql.org/message-id/CAHGQGwEtJCeHUB6KzaiJ6ndvx6EFsidTGnuLwJ1itwVH0EJTOA@mail.gmail.com
+                    if snapshot_id is None and not request.json.get('recovery', False):
+                        snapshot_id = connection.execute('SELECT pg_export_snapshot();').scalar()
+        return_now = False
+        return result, invalidated, flush, first_txn, snapshot_id, restart, xmin, return_now
 
-    indexing_update_infos = []
-    if invalidated and not dry_run:
+
+    def _run_indexing(
```

Comment:
move this function somewhere else, not in the middle of view method.   Ah, I see this was refactored from the old code into a method, it's not strictly new.